app/src/main/java/org/mozilla/focus/search/MultiselectSearchEngineListPreference.kt

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.search
+
+import android.app.Activity
+import android.content.Context
+import android.support.v7.preference.PreferenceViewHolder
+import android.view.ContextThemeWrapper
+import android.util.AttributeSet
+import android.widget.CompoundButton
+import org.mozilla.focus.R
+
+import java.util.HashSet
+
+class MultiselectSearchEngineListPreference(context: Context, attrs: AttributeSet) :
+        SearchEngineListPreference(context, attrs) {
+
+    override val itemResId: Int
+        get() = R.layout.search_engine_checkbox_button
+
+    val checkedEngineIds: Set<String>
+        get() {
+            val engineIdSet = HashSet<String>()
+
+            for (i in 0 until searchEngineGroup!!.childCount) {
+                val engineButton = searchEngineGroup!!.getChildAt(i) as CompoundButton
+                if (engineButton.isChecked) {
+                    engineIdSet.add(engineButton.tag as String)
+                }
+            }
+            return engineIdSet
+        }
+
+    override fun onBindViewHolder(holder: PreferenceViewHolder?) {
+        super.onBindViewHolder(holder)
+        this.bindEngineCheckboxesToMenu()
+    }
+
+    override fun updateDefaultItem(defaultButton: CompoundButton) {
+        defaultButton.isClickable = false
+        // Showing the default engine as disabled requires a StateListDrawable, but since there
+        // is no state_clickable and state_enabled seems to require a default drawable state,
+        // use state_activated instead to designate the default search engine.
+        defaultButton.isActivated = true
+    }
+
+    // Whenever an engine is checked or unchecked, we notify the menu
+    private fun bindEngineCheckboxesToMenu() {
+        for (i in 0 until searchEngineGroup!!.childCount) {
+            val engineButton = searchEngineGroup!!.getChildAt(i) as CompoundButton
+            engineButton.setOnCheckedChangeListener { _, _ ->
+                val context = context
+                if (context is ContextThemeWrapper) {
```

Comment:
I feel like this should be a global utility extension function on Context. We've repeated this pattern a lot in the code. Maybe Context.asActivity()?

Reply:
I found 2 such patterns. But how will the function actually work?
I am unable to understand how the utility function will handle the casting and after returning also execute the needed function call on the casted context

Reply:
Here's the simplest version. Put this utility function in a util class and use it.

`fun Context.asActivity() = (this as? ContextThemeWrapper)?.baseContext as? Activity ?: this as? Activity`

`context?.asActivity()?.invalidateOptionsMenu()`

Reply:
Using https://kotlinlang.org/docs/reference/typecasts.html#safe-nullable-cast-operator